Add MOOSE language server using WASP's LSP API

brandonlangley opened the pull request:
Add MOOSE language server using WASP's LSP API

refs #22766

Commit:
Add MOOSE language server using WASP's LSP API

refs #22766

brandonlangley commented:
@permcody -

This adds the initial `MooseServer` language server class which extends the `wasp::lsp::ServerImpl` abstract class.

Any `wasp::lsp::ServerImpl` child class meets the protocol by implementing these server-specific virtual methods:

```cpp
    /** parse document for diagnostics - to be implemented on derived servers
     * @param diagnosticsList - data array of diagnostics data objects to fill
     * @return - true if completed successfully - does not indicate parse fail
     */
    virtual bool parseDocumentForDiagnostics( DataArray  & diagnosticsList ) = 0;

    /** update document text changes - to be implemented on derived servers
     * @param replacement_text - text to be replaced over the provided range
     * @param start_line - starting replace line number ( zero-based )
     * @param start_character - starting replace column number ( zero-based )
     * @param end_line - ending replace line number ( zero-based )
     * @param end_character - ending replace column number ( zero-based )
     * @param range_length - length of replace range - server specific
     * @return - true if the document text was updated successfully
     */
    virtual bool updateDocumentTextChanges( const std::string & replacement_text ,
                                                  int           start_line       ,
                                                  int           start_character  ,
                                                  int           end_line         ,
                                                  int           end_character    ,
                                                  int           range_length     ) = 0;

    /** gather document completion items - to be implemented on derived servers
     * @param completionItems - data array of completion item objects to fill
     * @param is_incomplete - flag indicating if the completions are complete
     * @param line - line to be used for completions gathering logic
     * @param character - column to be used for completions gathering logic
     * @return - true if the gathering of items completed successfully
     */
    virtual bool gatherDocumentCompletionItems( DataArray & completionItems  ,
                                                bool      & is_incomplete    ,
                                                int         line             ,
                                                int         character        ) = 0;

    /** gather definition locations - to be implemented on derived servers
     * @param definitionLocations - data array of locations objects to fill
     * @param line - line to be used for locations gathering logic
     * @param character - column to be used for locations gathering logic
     * @return - true if the gathering of locations completed successfully
     */
    virtual bool gatherDocumentDefinitionLocations( DataArray & definitionLocations ,
                                                    int         line                ,
                                                    int         character           ) = 0;

    /** gather references locations - to be implemented on derived servers
     * @param referencesLocations - data array of locations objects to fill
     * @param line - line to be used for locations gathering logic
     * @param character - column to be used for locations gathering logic
     * @param include_declaration - flag indicating declaration inclusion
     * @return - true if the gathering of locations completed successfully
     */
    virtual bool gatherDocumentReferencesLocations( DataArray & referencesLocations ,
                                                    int         line                ,
                                                    int         character           ,
                                                    bool        include_declaration ) = 0;

    /** gather formatting text edits - to be implemented on derived servers
     * @param formattingTextEdits - data array of text edit objects to fill
     * @param start_line - starting line to be used for formatting logic
     * @param start_character - starting column to be used for formatting logic
     * @param end_line - ending line to be used for formatting logic
     * @param end_character - ending column to be used for formatting logic
     * @param tab_size - value of the size of a tab in spaces for formatting
     * @param insert_spaces - flag indicating whether to use spaces for tabs
     * @return - true if the gathering of text edits completed successfully
     */
    virtual bool gatherDocumentFormattingTextEdits( DataArray & formattingTextEdits ,
                                                    int         start_line          ,
                                                    int         start_character     ,
                                                    int         end_line            ,
                                                    int         end_character       ,
                                                    int         tab_size            ,
                                                    bool        insert_spaces       ) = 0;

    /** gather document symbols - to be implemented on derived servers
     * @param documentSymbols - data array of symbols data objects to fill
     * @return - true if the gathering of symbols completed successfully
     */
    virtual bool gatherDocumentSymbols( DataArray & documentSymbols ) = 0;
```

- - - - -

Right now, the `MooseServer` has a complete implementation for these three `wasp::lsp::ServerImpl` methods:

 - `bool parseDocumentForDiagnostics()`
 - `bool updateDocumentTextChanges()`
 - `bool gatherDocumentSymbols()`

The `parseDocumentForDiagnostics()` method creates a new app from the `AppFactory` with these parameters:

```cpp
    app_params.set<std::vector<std::string>>("input_file") = {parse_file_path};
    app_params.set<bool>("check_input")                    = true;
    app_params.set<bool>("error_unused")                   = true;
    app_params.set<bool>("error")                          = true;
    app_params.set<std::string>("color")                   = "off";
    app_params.set<bool>("disable_perf_graph_live")        = true;
    app_params.set<bool>("language_server")                = true;
```

It enables `Moose::_throw_on_error`, runs the app, and catches diagnostics thrown by the parse or check-input.

The `gatherDocumentSymbols()` method then walks the HIT tree and builds the protocol's hierarchical symbols.

- - - - -

This language server is started with the `--language-server` command line flag that will work for any MOOSE app.

The `MooseServer` been hooked up to the NEAMS Workbench and this workflow is currently working as expected:

 - starting and initializing the server
 - opening a MOOSE input file
 - receiving and showing diagnostics
 - modifying the input file contents
 - displaying the updated diagnostics
 - receiving the document symbols
 - building hierarchical navigation tree
 - shutting the server down and exiting

- - - - -

These other `wasp::lsp::ServerImpl` methods exist in the `MooseServer` but are basically no-ops at the moment:

 - `bool gatherDocumentCompletionItems()`
 - `bool gatherDocumentDefinitionLocations()`
 - `bool gatherDocumentReferencesLocations()`
 - `bool gatherDocumentFormattingTextEdits()`

So autocompletion, goto-definition, find-references, and input-formatting are currently unavailable in Workbench.

But, I have started working on the `gatherDocumentCompletionItems()` implementation  to enable autocompletion.

- - - - -

Since this uses WASP's LSP API, the branch from this PR is based on the WASP integration branch from https://github.com/idaholab/moose/pull/22018.

This PR should not be merged until after https://github.com/idaholab/moose/pull/22018 is reviewed and merged, so I made this one a "draft" until then.

dschwen commented:
What's the performance for this? How does it handle input file modification. Does that require a restart?

brandonlangley commented:
> What's the performance for this? How does it handle input file modification. Does that require a restart?

@dschwen -

When the input document is first opened, this communication takes place:

```
language server is started via the `moose-opt --language-server' flag

Client --> initialize-request              --> MooseServer

Client <-- initialize-response             <-- MooseServer

Client --> initialized-notification        --> MooseServer

Client --> didOpen-notification            --> MooseServer

Client <-- publishDiagnostics-notification <-- MooseServer

Client --> documentSymbol-request          --> MooseServer

Client <-- documentSymbol-response         <-- MooseServer
```

Then that same MOOSE language server stays active and persists until the input is closed or client is exited.

So the server is not required to restart and suffer any startup / handshake time when an input file is modified.

For all input changes after the document is opened, a `didChange` rather than a `didOpen` notification is sent.

There are two `TextDocumentSyncKind` types in the protocol:
```
 - Full         // Documents are synced by always sending the full content of the document.
 - Incremental  // Documents are synced by sending the full content on open.
                // After that only incremental updates to the document are sent.
```

The `initialize-response` sent from the MooseSever to the client on startup contains the server capabilities.

For the MooseServer, this `capabilities/textDocumentSync/change` value is communicated to be `Full` type.

So currently the entire document is sent on `didChange` rather than specific range of changes in the document.

Then that new document is parsed and the check-input is ran catching any errors and sending as diagnostics.

The NEAMS Workbench uses a timer so this is not sent for every keystroke - but only when typing has stopped.

It also doesn't build backlog of `didChange` notifications and doesn't send another until diagnostics come back.

Using the MooseServer in Workbench to edit a 281 line input for diagnostics and hierarchical symbols is snappy.

The input-check in the `parseDocumentForDiagnostics()` method does currently check mesh generation errors.

So if that mesh is large, that part can cause a delay in receiving the diagnostics for a `didChange` notification.

But, the actual LSP communication between Workbench and the MooseServer while editing is quite responsive.

dschwen commented:
> For all input changes after the document is opened, a `didChange` rather than a `didOpen` notification is sent.

I understand that part, I'm just wondering how you handle the previously constructed objects when you reparse the file. I'm assuming everything gets destroyed and then rebuilt by the factory? Are you tearing down the MooseApp object? I should probably read the code :-D, but I'm on travel and my phone screen is small...

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+bool
+MooseServer::setup(MooseApp * moose_app)
+{
+  bool pass = true;
+
+  // cache a pointer to the application that set up this server instance
+
+  this->moose_app = moose_app;
+
+  this->is_setup = true;
+
+  return pass;
```
why not `return true;`, what is the purpose of the `pass` variable?

brandonlangley replied:
Good feedback - I just added a commit addressing this.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+  app_params.set<std::vector<std::string>>("input_file") = {parse_file_path};
+
+  app_params.set<bool>("check_input") = true;
+
```
please use empty lines sparingly to separate sections of code

brandonlangley replied:
Good feedback - I just added a commit addressing this.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+bool
+MooseServer::parseDocumentForDiagnostics(wasp::DataArray & diagnosticsList)
+{
+  if (!this->is_setup)
```
```suggestion
  if (_is_setup)
```
please use a leading `_` for member variables and avoid `this->` where it is not explicitly needed

brandonlangley replied:
Good feedback - I just added a commit addressing this.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+
+      // according to the protocol line and column number are zero based
+
+      pass &= wasp::lsp::buildDiagnosticObject(*diagnostic,
```
does it make sense to continue if this step failed?

brandonlangley replied:
This is a good question - that method would return false if some parameter was illegal according to the protocol.

For example, if a line number is negative, then an LSP error with an appropriate message would be communicated.

But the client, depending on implementation, could display the input diagnostics along with handling the LSP error.

So if this fails for a diagnostic, I don't think it's a bad idea to continue walking the input errors and adding the rest.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+  return pass;
+}
+
+bool
+MooseServer::gatherDocumentCompletionItems(wasp::DataArray & completionItems,
+                                           bool & is_incomplete,
+                                           int line,
+                                           int character)
+{
+  (void)completionItems; // suppress unused variable warning
+  (void)is_incomplete;   // suppress unused variable warning
+  (void)line;            // suppress unused variable warning
+  (void)character;       // suppress unused variable warning
```
```suggestion
bool
MooseServer::gatherDocumentCompletionItems(wasp::DataArray & /*completionItems*/,
                                           bool & /*is_incomplete*/,
                                           int /*line*/,
                                           int /*character*/)
{
```

brandonlangley replied:
Good feedback - I just added a commit addressing this.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+                                               int tab_size,
+                                               bool insert_spaces)
+{
+  (void)formattingTextEdits; // suppress unused variable warning
```
see above

brandonlangley replied:
Good feedback - I just added a commit addressing this.

dschwen commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+
+  wasp::HITNodeView view_root = this->check_app->parser()._root->getNodeView();
+
+  for (std::size_t i = 0, count = view_root.child_count(); i < count; i++)
```
```suggestion
  for (const auto i: make_range(view_root.child_count())
```

is the new way of setting up loop heders

brandonlangley replied:
Good feedback - I just added a commit addressing this.

Commit:
Address MooseServer feedback from code review

refs #22766

brandonlangley commented on framework/src/base/MooseServer.C:
```diff
@@ -0,0 +1,529 @@
+    return false;
+  }
+
+  bool pass = true;
```
Good feedback - I just added a commit addressing this.

brandonlangley commented:
@dschwen -

Thanks a lot for the comments - I just added a [commit](https://github.com/idaholab/moose/pull/22768/commits/0af51357a2f096286dff0e00c55d168d698dbe88) to address all of your feedback.

Also, earlier I said:

> Since this uses WASP's LSP API, the branch from this PR is based on the WASP integration branch from #22018.
>
> This PR should not be merged until after #22018 is reviewed and merged, so I made this one a "draft" until then.

But since the first [commit](https://github.com/idaholab/moose/pull/22768/commits/cc19553d11d9762da1d00e29c23c95810c5a6cb8) on this PR is the one commit from #22018, reviewing this would include the other.

So I went ahead and moved this out of "draft" since, if preferred, this could be merged as a whole instead of #22018.

Again, this PR currently includes implementations of these LSP features in the MooseServer:

 - input diagnostics feedback
 - hierarchical tree navigation

Implementations for other LSP features such as the following could be added on a future PR:

 - autocompletion assistance
 - goto-definition assistance
 - find-references assistance
 - input formatting assistance

This would be done by filling in the logic for these (currently no-op) MooseServer methods:

 - `bool gatherDocumentCompletionItems()`
 - `bool gatherDocumentDefinitionLocations()`
 - `bool gatherDocumentReferencesLocations()`
 - `bool gatherDocumentFormattingTextEdits()`

Commit:
Remove disk IO from MooseServer input processing

refs #22766

brandonlangley commented:
I rebased this branch's commits on top of the six commits from the dependency https://github.com/idaholab/moose/pull/22018 `wasp_hit_updates_22015` branch. 

I plan to drop all of those base commits from here after after https://github.com/idaholab/moose/pull/22018 gets merged and then just rebase this branch on next.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1517119/) on 6592357 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22768/site)

This comment will be updated on new commits.

loganharbour commented on framework/include/base/MooseServer.h:
```diff
@@ -0,0 +1,172 @@
+   * @param moose_app - pointer to moose app calling this set up method
+   * @return - true if server setup was successful and everything was set
+   */
+  bool setup(MooseApp * moose_app);
```
Why is this method necessary? It looks you should store a reference to the MooseApp instead as a member variable, and pass the app in the constructor of the MooseServer.

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -944,6 +960,12 @@ MooseApp::setupOptions()
       _input_filenames = getParam<std::vector<std::string>>("input_file");
 
+    // Reset output to the buffer what was cached before it was turned it off
+    if (!Moose::out.rdbuf() && _output_buffer_cache)
+    {
+      Moose::out.rdbuf(_output_buffer_cache);
+    }
+
```
```suggestion
    if (!Moose::out.rdbuf() && _output_buffer_cache)
      Moose::out.rdbuf(_output_buffer_cache);

```

loganharbour reviewed (changes requested):
Minor changes. Few general suggestions that I try to follow:

- If something shouldn't ever be nullptr and you're passing it in a function, use a reference, not a pointer
- If not using a shared pointer works, don't use one. If you need a pointer so that you can construct something later, use unique_ptr. If you _actually_ need to share something, use a shared_ptr.

brandonlangley committed:
Update MooseServer from code review suggestions

refs #22766

brandonlangley commented:
@loganharbour -

Thanks a lot for the review and feedback - I just added a [commit](https://github.com/idaholab/moose/pull/22768/commits/d995fa34c67d3a6d8f2b6955b26c1bc9e2d4f884) with all your suggestions except the two listed below.

And I am about to work on the `MooseServer` unit testing now, so I plan to have one more commit with that added soon.

1. > `_connection = std::make_shared<wasp::lsp::IOStreamConnection>(this);`
    >
    > Why does the connection need to be a shared_ptr? If it doesn't actually need to be shared, it should be a unique_ptr and more likely doesn't even need to be a smart pointer and can just be an object.
    
    The issue is `MooseServer` extends `wasp::lsp::ServerImpl` which implements common language server functions.
    
    The servers derived from `wasp::lsp::ServerImpl` just implement the virtual methods with language specific logic.
    
    That base `ServerImpl` class defines the `Connection` as needing to be stored as a `shared_ptr` for a few reasons.
    
    While `wasp::lsp::IOStreamConnection` is used here, there are other `wasp::lsp::Connection` implementations too.
    
    For example, `wasp::lsp::ThreadConnection` facilitates  client and language server communication using this setup:
    
    - launch a server on separate thread to start running
    - get server read / write connection as a shared_ptr
    - pass that server connection shared_ptr to the client
    - server and client share connection to communicate
    
    There are also unit testing scenarios in WASP that share a connection from a server for direct reading and writing.
    
    This allows testing of the server capabilities and low level LSP features directly without having to introduce a client.
    
    In the future, there may be cases of wanting `MooseServer` to use `wasp::lsp::ThreadConnection` for some reason.
    
    One implemented server has its `wasp::lsp::Connection` type templated and works with various connection types.
    
    But I don't think we need to template the connection type for the `MooseServer` yet - unless it is needed for testing.

1. > `std::shared_ptr<MooseApp> _check_app;`
    >
    > Should probably be unique_ptr.
    
    I thought a `MooseApp` had to be created with the `AppFactory` since its constructor is protected with this comment:
    
    ```cpp
    /// Constructor is protected so that this object is constructed through the AppFactory object
    MooseApp(InputParameters parameters);
    ```
    
    I may be misunderstanding, but these only `AppFactory` methods to create a `MooseApp` both return a `shared_ptr`:
    
    ```cpp
    using MooseAppPtr = std::shared_ptr<MooseApp>;
    
    . . .
    
    MooseAppPtr createShared(const std::string & app_type,
                             const std::string & name,
                             InputParameters parameters,
                             MPI_Comm COMM_WORLD_IN);
    . . .
    
    MooseAppPtr createShared(const std::string & app_type,
                             const std::string & name,
                             InputParameters parameters,
                             MPI_Comm COMM_WORLD_IN);
    ```
    
    While a conversion from a `unique_ptr` to a `shared_ptr` is possible, it is not from a `shared_ptr` to a `unique_ptr`.
    
    So I couldn't find a straight-forward way of changing this `MooseApp` to be a `unique_ptr` instead of a `shared_ptr`.
    
    If `_check_app` really needs to be a `unique_ptr`, do you have any suggestions for a better way to implement this?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1523615/) on d995fa3 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1523615/) on d995fa3 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

brandonlangley committed:
Unit test all current MooseServer capabilities

refs #22766

brandonlangley commented:
@loganharbour -

I just added a [commit](https://github.com/idaholab/moose/pull/22768/commits/1946a301e924e4fc32f75de54b2bc1bd1106cae9) that has unit testing for all of the current MooseServer capabilities.

 - initialize capabilities
 - document open
 - input diagnostics
 - document change
 - hierarchical symbols

I plan to implement autocompletion,  goto definition, find references, etc. on a separate PR.

brandonlangley committed:
Split MooseServer unit test into separate cases

refs #22766

brandonlangley commented:
@loganharbour -

I originally had the MooseServer unit test as one long test case since the server state needs to persist across calls.

For example, you have make the language server protocol initialize() call before you can do a documentOpen() call.

Similarly, you have to call documentOpen() before you can call documentChange() or request documentSymbols().

And if shared data is modified in one test case, those changes are not maintained for subsequent cases by default.

But I didn't like having the unit test as just one long case that marched through all of the language server features.

So I found a way with Gtest's SetUpTestCase() / TearDownTestCase() calls to persist modified data between cases.

And I just added one last commit to this PR that cleaned up the test by splitting it into cases based on capabilities.

But if this setup of reusing the MooseServer across tests causes any issues, that final commit can just be dropped.

loganharbour commented on framework/include/base/MooseServer.h:
```diff
@@ -0,0 +1,172 @@
+  /**
+   * @brief _check_app - application created to check input and access parser
+   */
+  std::shared_ptr<MooseApp> _check_app;
```
In the future please respond to comments where they are made. It makes it easier to review as the context is immediately in view.

Additionally, please avoid resolving comments unless there is very explicit agreement. Something like a suggested change that is requested and then exactly made

loganharbour commented on framework/src/parser/Parser.C:
```diff
@@ -637,10 +637,14 @@ Parser::parse(const std::vector<std::string> & input_filenames)
-    std::ifstream f(input_filename);
-    std::string input((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
+    // Parse the input text string if non-empty, otherwise read file from disk
+    std::string input(input_text);
+    if (input.empty())
+    {
+      MooseUtils::checkFileReadable(input_filename, true);
+      std::ifstream f(input_filename);
+      input = std::string((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());
+    }
```
This logic is weird to me. I think that if you provide multiple filenames... and an input text, it would read the input text multiple times... yeah?

brandonlangley replied:
Yes, this is true and admittedly a bit weird.

The idea here is that the `_input_text` parameter is private and hidden in the MooseApp.

And this option is only used by the MooseServer to remove disk IO from input processing.

The MooseServer also only sets a single `input_filename` when it sets the `_input_text`.

So there would never be more than one `input_filename` when `input_text` is also used.

But this not a good assumption for future-proofing using the private `_input_text` option.

I can add this check to the top of the `parse()` method to prevent that case you describe.

```cpp
if (!input_text.empty() && input_filenames.size() != 1)
  mooseError("If 'input_text' is provided, then 'input_filenames' must contain a single filename");
```

There will currently be no way to hit that error in practice, but it will guard it for the future.

Is this an okay check to stop the unintended case of reading `input_text` multiple times?

loganharbour replied:
This isn't pretty, but I don't think it's a significant concern

loganharbour commented on unit/src/MooseServerTest.C:
```diff
@@ -0,0 +1,531 @@
+    }
+  }
+
+  // create moose_unit_app and moose_server to persist for reuse between tests
+  static void SetUpTestCase()
+  {
+    moose_unit_app = AppFactory::createAppShared("MooseUnitApp", 0, nullptr);
+    moose_server = std::make_unique<MooseServer>(*moose_unit_app);
+  }
+
+  // delete moose_unit_app and moose_server after all test runs have completed
+  static void TearDownTestCase()
+  {
+    moose_unit_app.reset();
+    moose_server.reset();
+  }
+
+  // statically declare moose_unit_app and moose_server for reuse across tests
+  static std::shared_ptr<MooseApp> moose_unit_app;
+  static std::unique_ptr<MooseServer> moose_server;
```
Is static the gest-approved way to do this? It doesn't seem thread safe, but... I guess this isn't threaded

brandonlangley replied:
Yes, the gtest `void SetUpTestCase()` and `void TearDownTestCase()` have to be defined as static methods.

And these static methods are only allowed to operate on static member variable data.

So static is not only the recommended way to do data setup to be reused across all test cases, it is required.

loganharbour reviewed (changes requested):
Minor questions/comments.

Have you looked at the coverage diff: https://mooseframework.inl.gov/docs/PRs/22768/coverage_diff/framework/ to see if you can add any extra test coverage?

brandonlangley commented on framework/src/base/MooseApp.C:
```diff
@@ -129,6 +130,13 @@ MooseApp::validParams()
       false,
       "Ignore input file and build a minimal application with Transient executioner.");
 
+#ifdef WASP_ENABLED
+  params.addCommandLineParam<bool>(
+      "language_server",
+      "--language-server",
+      "Starts a process to communicate with development tools using the language server protocol");
+#endif
```
Our Workbench integration plan is to grep the `app-opt --help` output to see if the language server is available.

 - if `--language-server` is in the available options, then we will start and use the MOOSE language server
 
 - if it doesn't show up in the options, we will fall back to the old `--definition` schema based integration

Otherwise, Workbench will have to blindly try starting the language server then wait a millisecond for it to error.

After which that subprocess would be killed and Workbench would fall back to the `--definition` schema route.

We would prefer to efficiently check which integration route to take from the options in `app-opt --help` output.

But if there's a strong preference to always show the option even when it's not available, I can make this change.

loganharbour replied:
This is fine.

brandonlangley committed:
Increase test coverage percent for MooseServer

refs #22766

brandonlangley committed:
Check only one filename if parsing input string

refs #22766

brandonlangley commented:
@loganharbour -

This is back out of draft status and ready to be reviewed again.

I [added](https://github.com/idaholab/moose/pull/22768/commits/c87cc5b4eb5ecd48f838ce84c9596a21ebbb5d4c) a check that only one filename is provided when parsing an input string.

I also [increased](https://mooseframework.inl.gov/docs/PRs/22768/coverage_diff/framework/) the test coverage for new code from 70.19% to 92.65%.

I left comments for the two remaining questions you had above.

Just let me know if you have any questions.

loganharbour reviewed (approved):
I would like for us to be able to test the actual server itself, but we can work on that in the future.
